src/taoensso/faraday.clj

```diff
@@ -1210,6 +1232,111 @@
                   (mapv (partial write-request action) items))))
             requests))))))
 
+(defn- cond-check
+  [table prim-kvs cond-expr
+   {:keys [expr-attr-names expr-attr-vals return]
+    :or   {return :none}}]
+  (doto-cond [g (ConditionCheck.)]
+    :always         (.setTableName    (name table))
+    :always         (.setKey          (clj-item->db-item prim-kvs))
+    :always         (.setConditionExpression cond-expr)
+    expr-attr-names (.withExpressionAttributeNames expr-attr-names)
+    expr-attr-vals  (.withExpressionAttributeValues
+                      (clj->db-expr-vals-map expr-attr-vals))
+    return          (.setReturnValuesOnConditionCheckFailure (utils/enum g))))
+
+(defn- put
+  [table item {:keys [cond-expr expr-attr-names expr-attr-vals return]
+               :or {return :none}}]
+  (doto-cond [g (Put.)]
+    :always         (.setTableName (name table))
+    :always         (.setItem (clj-item->db-item item))
+    cond-expr       (.setConditionExpression cond-expr)
+    expr-attr-names (.withExpressionAttributeNames expr-attr-names)
+    expr-attr-vals  (.withExpressionAttributeValues
+                     (clj->db-expr-vals-map expr-attr-vals))
+    return          (.setReturnValuesOnConditionCheckFailure (utils/enum g))))
+
+(defn- delete
+  [table-name prim-kvs {:keys [cond-expr expr-attr-names expr-attr-vals return]
+                   :or {return :none}}]
+  (doto-cond [g (Delete.)]
+    :always (.setTableName (name table-name))
+    :always (.setKey (clj-item->db-item prim-kvs))
+    cond-expr       (.setConditionExpression cond-expr)
+    expr-attr-names (.withExpressionAttributeNames expr-attr-names)
+    expr-attr-vals  (.withExpressionAttributeValues
+                     (clj->db-expr-vals-map expr-attr-vals))
+    return          (.setReturnValuesOnConditionCheckFailure (utils/enum g))))
+
+(defn- update
+  [table-name prim-kvs update-expr
+   {:keys [cond-expr expr-attr-names expr-attr-vals return]
+    :or {return :none}}]
+  (doto-cond [g (Update.)]
+    :always         (.setTableName (name table-name))
+    :always         (.setKey (clj-item->db-item prim-kvs))
+    :always         (.setUpdateExpression update-expr)
+    cond-expr       (.setConditionExpression cond-expr)
+    expr-attr-names (.withExpressionAttributeNames expr-attr-names)
+    expr-attr-vals  (.withExpressionAttributeValues
+                     (clj->db-expr-vals-map expr-attr-vals))
+    return          (.setReturnValuesOnConditionCheckFailure (utils/enum g))))
+
+(defmulti ^:private transact-write-item (comp first keys))
```

Comment:
Maybe I'm being dense here, but shouldn't we have vector pairs here rather than a map to represent each item?

It's not as if you can have multiple updates in each item, like:

```clj
{:put {...}
 :delete {...}}
```

In each item you have only one operation allowed.

If we look at the Faraday `query` syntax as an example, we represent the query as a pair like:

```
[<comparison-operator> <value>]
``` 

e.g. 

```
[:eq "abc]
```

Yes we use maps for `batch-write-item`, but this is because any/all operations we include as keys in the map will be applied.

So I think this

```clj
(defmulti ^:private transact-write-item (comp first keys))
```

should become this:
```clj
(defmulti ^:private transact-write-item first)
```
and this in each defmethod:
```clj
[{{:keys [table-name prim-kvs cond-expr] :as request} :cond-check}]
```
should become this:
```clj
[[_ {:keys [table-name prim-kvs cond-expr] :as request}]]
```

What do you think?

Reply:
I'd probably also add an assertion that the `count` of the given arg is 2.